urld2fb1d2b05_thnaeff_Util_tgz-pJ8-ch_thn_util_tests_TreeTestJ8/src/ch/thn/util/NumberUtil.java

```diff
@@ -0,0 +1,612 @@
+	 * 
+	 * @param min
+	 * @param max
+	 * @return
+	 */
+	public static Number generateRandomNumber(Number min, Number max) {
+		return generateRandomNumbers(1, min, max).get(0);
+	}
+
+	/**
+	 * Generates random numbers. The type of the resulting numbers is the type of the
+	 * given min or max value, whichever has the higher "precision"
+	 * (short -> int -> long -> float -> double)
+	 * 
+	 * @param count
+	 * @param min
+	 * @param max
+	 * @return
+	 */
+	public static ArrayList<Number> generateRandomNumbers(int count, Number min, Number max) {
+		Random random = new Random();
+		ArrayList<Number> numbers = new ArrayList<>(count);
+
+		int type = getOperationType(min, max);
+
+		if (type == SHORT) {
+			int randomMax = max.intValue() - min.intValue() + 1;	//Including the maximum number (from min to max)
+			for (int i = 0; i < count; i++) {
+				numbers.add(random.nextInt(randomMax) + min.intValue());
+			}
+		} else if (type == INTEGER) {
+			int randomMax = max.intValue() - min.intValue() + 1;	//Including the maximum number (from min to max)
+			for (int i = 0; i < count; i++) {
+				numbers.add(random.nextInt(randomMax) + min.intValue());
+			}
+		} else if (type == LONG) {
+			long randomMax = max.longValue() - min.longValue() + 1l;	//Including the maximum number (from min to max)
+			for (int i = 0; i < count; i++) {
+				numbers.add(random.nextLong() * randomMax + min.longValue());
+			}
+		} else if (type == FLOAT) {
+			float randomMax = max.floatValue() - min.floatValue();	//Excluding the maximum number (between min and max)
+			for (int i = 0; i < count; i++) {
+				numbers.add(random.nextFloat() * randomMax + min.floatValue());
+			}
+		} else if (type == DOUBLE) {
+			double randomMax = max.doubleValue() - min.doubleValue();	//Excluding the maximum number (between min and max)
+			for (int i = 0; i < count; i++) {
+				numbers.add(random.nextDouble() * randomMax + min.doubleValue());
+			}
+		}
+
+		return numbers;
+	}
+
+	/**
+	 * Casts the given number to the given class.<br>
+	 * Valid classes are all the number or primitive classes:<br>
+	 * - short.class, Short.class, Short.TYPE<br>
+	 * - int.class, Integer.class, Integer.TYPE<br>
+	 * - double.class, Double.class, Double.TYPE<br>
+	 * etc.
+	 * 
+	 * @param c
+	 * @param number
+	 * @return
+	 */
+	public static Number castTo(Class<?> c, Number number) {
+
+		String primitiveString = null;
+
+		if (ReflectionUtil.primitiveTypeMapper.containsKey(c)) {
+			//e.g. Double.TYPE or double.class
+			primitiveString = (String)ReflectionUtil.primitiveTypeMapper.get(c)[0];
+		} else if (ReflectionUtil.primitiveClassMapper.containsKey(c)) {
+			//e.g. Double.class
+			primitiveString = (String)ReflectionUtil.primitiveClassMapper.get(c)[0];
+		}
+
+		if (primitiveString == null) {
+			throw new NumberUtilError("Invalid number cast. " + c.getCanonicalName() + " is not a valid class.");
+		}
+
+		switch (primitiveString) {
+		case "int":
+			return number.intValue();
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem:**
 This line of code lacks validation when processing input data through the following parameter: &#39;number&#39; (index: 1 | type: Number). The parameter is exposed to external callers, because its enclosing class and method are publicly accessible. This means that upstream validation, if it exists, can be bypassed. Other validated parameters: &#39;c&#39;. The same parameter type is validated here for example: urld2fb1d2b05_thnaeff_Util_tgz-pJ8-ch_thn_util_tests_TreeTestJ8/src/ch/thn/util/NumberUtil.java:488. Malicious, malformed, or unbounded inputs can cause unexpected runtime behavior or crashes, and can slow performance.

**Fix:**
 Add checks to ensure the validity of the parameter&#39;s value, such as testing it for nullness (for example, using the @NonNull annotation as [described in the Lombok library](https://projectlombok.org/features/NonNull)), emptiness, or equality. Or to prevent direct calls to it, [reduce the method&#39;s visibility](https://docs.oracle.com/javase/tutorial/java/javaOO/accesscontrol.html).

 Learn more about potential threats and guidance from the [Common Weakness Enumeration website](https://cwe.mitre.org/data/definitions/693.html) and the [OWASP Cheat Sheet series](https://cheatsheetseries.owasp.org/cheatsheets/Input_Validation_Cheat_Sheet.html).